# 01. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Import dataset

In [2]:
path = r'C:\Users\Logan\OneDrive\Desktop\Instacart Basket Analysis'
path

'C:\\Users\\Logan\\OneDrive\\Desktop\\Instacart Basket Analysis'

In [4]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', '4.8_ords_prods_merge_grouping data'))

In [5]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    Busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy  Regularly busy        Average orders         10   
2  Regularly busy  Regularly busy        Average orders         10   
3      Least busy      Least busy        Average orders         10   
4      Least busy      Least busy        Average orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [6]:
ords_prods_merge.shape

(32404859, 20)

# 03. Tasks

In [10]:
# aggregate the mean of the "order_number" column grouped by "department_id" for entire dataframe
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

Analyze the results of average order by department id:

- Department 21 has the highest average order number
- Compared to the subset, average orders are higher in the subset for certain department_ids or lower as it was a subset of the data. For example, Department_id 19 averages 17.1 orders vs 19.3 in the subset of data.

In [12]:
# Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer).

ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean','min','max','count']})

prices                        
                       mean  min      max     count
loyalty_flag                                       
Loyal customer    10.386336  1.0  99999.0  10284093
New customer      13.294670  1.0  99999.0   6243990
Regular customer  12.495717  1.0  99999.0  15876776

Loyal customers spend on average less based on product price

In [14]:
# Team now wants to target different types of spenders
# Create spending flag for each user based on criteria
# 1. if the mean of prices of products purchased by a user is lower than 10
#    then flag them as a "low spender"
# 2. if the mean of prices of products purchased by a user is higher than or equal to 10
#    then flag them as a "High spender"

# split the data into groups based on user_id
# apply the transform() on prices to generate average price by user
# create a new column, mean_price to place results
ords_prods_merge['mean_price'] = ords_prods_merge.groupby('user_id')['prices'].transform(np.mean)

In [16]:
ords_prods_merge[['user_id', 'mean_price']].head(1)

user_id  mean_price
0        1    6.367797

In [17]:
# use results and derive new column spending_flag

ords_prods_merge.loc[ords_prods_merge['mean_price'] < 10, 'spending_flag'] = 'Low spender'

ords_prods_merge.loc[ords_prods_merge['mean_price'] >= 10, 'spending_flag'] = 'High spender'

In [19]:
# display more rows
pd.set_option('display.max_rows', 100)

In [21]:
# QA check flag
ords_prods_merge[['user_id','mean_price', 'spending_flag']].head(100)

user_id  mean_price spending_flag
0         1    6.367797   Low spender
1         1    6.367797   Low spender
2         1    6.367797   Low spender
3         1    6.367797   Low spender
4         1    6.367797   Low spender
5         1    6.367797   Low spender
6         1    6.367797   Low spender
7         1    6.367797   Low spender
8         1    6.367797   Low spender
9         1    6.367797   Low spender
10       15    3.980556   Low spender
11       15    3.980556   Low spender
12       15    3.980556   Low spender
13       15    3.980556   Low spender
14       15    3.980556   Low spender
15       19    8.372059   Low spender
16       19    8.372059   Low spender
17       19    8.372059   Low spender
18       21    7.012683   Low spender
19       31    8.137793   Low spender
20       31    8.137793   Low spender
21       43    7.444898   Low spender
22       43    7.444898   Low spender
23       52    6.356287   Low spender
24       52    6.356287   Low spender
25       52    6.356287   Low spender
26       52    6.356287   Low spender
27       52    6.356287   Low spender
28       52    6.356287   Low spender
29       52    6.356287   Low spender
30       52    6.356287   Low spender
31       52    6.356287   Low spender
32       52    6.356287   Low spender
33       52    6.356287   Low spender
34       52    6.356287   Low spender
35       52    6.356287   Low spender
36       52    6.356287   Low spender
37       67    7.656790   Low spender
38       67    7.656790   Low spender
39       67    7.656790   Low spender
40       67    7.656790   Low spender
41       67    7.656790   Low spender
42       67    7.656790   Low spender
43       67    7.656790   Low spender
44       67    7.656790   Low spender
45       67    7.656790   Low spender
46       67    7.656790   Low spender
47       67    7.656790   Low spender
48       67    7.656790   Low spender
49       67    7.656790   Low spender
50       67    7.656790   Low spender
51       67    7.656790   Low spender
52       67    7.656790   Low spender
53       67    7.656790   Low spender
54       67    7.656790   Low spender
55       67    7.656790   Low spender
56       81    8.165385   Low spender
57       81    8.165385   Low spender
58       82    6.441221   Low spender
59       82    6.441221   Low spender
60       82    6.441221   Low spender
61       82    6.441221   Low spender
62       82    6.441221   Low spender
63       82    6.441221   Low spender
64       82    6.441221   Low spender
65       82    6.441221   Low spender
66       82    6.441221   Low spender
67       98    8.028000   Low spender
68       98    8.028000   Low spender
69       98    8.028000   Low spender
70       98    8.028000   Low spender
71       98    8.028000   Low spender
72       98    8.028000   Low spender
73       98    8.028000   Low spender
74      109    5.850725   Low spender
75      120    9.385714   Low spender
76      120    9.385714   Low spender
77      185    7.715000   Low spender
78      195    7.149730   Low spender
79      195    7.149730   Low spender
80      195    7.149730   Low spender
81      195    7.149730   Low spender
82      195    7.149730   Low spender
83      195    7.149730   Low spender
84      195    7.149730   Low spender
85      195    7.149730   Low spender
86      195    7.149730   Low spender
87      195    7.149730   Low spender
88      195    7.149730   Low spender
89      195    7.149730   Low spender
90      222    6.862397   Low spender
91      222    6.862397   Low spender
92      290    7.285597   Low spender
93      290    7.285597   Low spender
94      331    7.705128   Low spender
95      360   10.006250  High spender
96      377    8.496552   Low spender
97      387    7.396610   Low spender
98      420    7.387805   Low spender
99      420    7.387805   Low spender

In [22]:
# Team wants to distinguish frequent versus non-frequent customers to send relevant notifications if they want to buy the same item again
# Create order frequency flag that makrs the regularity of a user's ordering behavior
    # according to the median in the "days_since_prior_order" column
# Criteria
# 1. if the median of “days_since_prior_order” is higher than 20, 
    #then the customer should be labeled a “Non-frequent customer.”
# 2. If the median is higher than 10 and lower than or equal to 20, 
    # then the customer should be labeled a “Regular customer.”
# 3. If the median is lower than or equal to 10, 
    # then the customer should be labeled a “Frequent customer.”
    
# split the data into groups based on user_id
# apply the transform() on days prior to generate median 
# create a new column, median_days_since_prior_order to place results
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform(np.median)

In [24]:
ords_prods_merge[['user_id', 'days_since_prior_order','median_days_since_prior_order']].head(11)

user_id  days_since_prior_order  median_days_since_prior_order
0         1                     NaN                           20.5
1         1                    15.0                           20.5
2         1                    21.0                           20.5
3         1                    29.0                           20.5
4         1                    28.0                           20.5
5         1                    19.0                           20.5
6         1                    20.0                           20.5
7         1                    14.0                           20.5
8         1                     0.0                           20.5
9         1                    30.0                           20.5
10       15                     7.0                           10.0

In [25]:
# Use results and derive new column order_frequency_flag

ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] <= 20) & (ords_prods_merge['median_days_since_prior_order'] > 10), 'order_frequency_flag'] = 'Regular customer'

ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'



In [27]:
# QA check order_frequency_flag
ords_prods_merge[['user_id', 'median_days_since_prior_order', 'order_frequency_flag']].head(100)

user_id  median_days_since_prior_order   order_frequency_flag
0         1                           20.5  Non-frequent customer
1         1                           20.5  Non-frequent customer
2         1                           20.5  Non-frequent customer
3         1                           20.5  Non-frequent customer
4         1                           20.5  Non-frequent customer
5         1                           20.5  Non-frequent customer
6         1                           20.5  Non-frequent customer
7         1                           20.5  Non-frequent customer
8         1                           20.5  Non-frequent customer
9         1                           20.5  Non-frequent customer
10       15                           10.0      Frequent customer
11       15                           10.0      Frequent customer
12       15                           10.0      Frequent customer
13       15                           10.0      Frequent customer
14       15                           10.0      Frequent customer
15       19                            6.0      Frequent customer
16       19                            6.0      Frequent customer
17       19                            6.0      Frequent customer
18       21                            7.5      Frequent customer
19       31                            6.0      Frequent customer
20       31                            6.0      Frequent customer
21       43                           10.0      Frequent customer
22       43                           10.0      Frequent customer
23       52                            8.0      Frequent customer
24       52                            8.0      Frequent customer
25       52                            8.0      Frequent customer
26       52                            8.0      Frequent customer
27       52                            8.0      Frequent customer
28       52                            8.0      Frequent customer
29       52                            8.0      Frequent customer
30       52                            8.0      Frequent customer
31       52                            8.0      Frequent customer
32       52                            8.0      Frequent customer
33       52                            8.0      Frequent customer
34       52                            8.0      Frequent customer
35       52                            8.0      Frequent customer
36       52                            8.0      Frequent customer
37       67                            7.0      Frequent customer
38       67                            7.0      Frequent customer
39       67                            7.0      Frequent customer
40       67                            7.0      Frequent customer
41       67                            7.0      Frequent customer
42       67                            7.0      Frequent customer
43       67                            7.0      Frequent customer
44       67                            7.0      Frequent customer
45       67                            7.0      Frequent customer
46       67                            7.0      Frequent customer
47       67                            7.0      Frequent customer
48       67                            7.0      Frequent customer
49       67                            7.0      Frequent customer
50       67                            7.0      Frequent customer
51       67                            7.0      Frequent customer
52       67                            7.0      Frequent customer
53       67                            7.0      Frequent customer
54       67                            7.0      Frequent customer
55       67                            7.0      Frequent customer
56       81                           30.0  Non-frequent customer
57       81                           30.0  Non-frequent customer
58       82                           11.5       Regular customer
59       82                           11.5  

In [29]:
# QA check frequency
ords_prods_merge['order_frequency_flag'].value_counts(dropna=False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [33]:
# View NaN results in order_frequency_flag
# new customer with 1 order history
ords_prods_merge[ords_prods_merge['order_frequency_flag'].isnull()]

order_id  user_id  order_number  order_day_of_week  \
13645692    895835   159838             1                  0   
17251990    895835   159838             1                  0   
17622767    895835   159838             1                  0   
24138593    895835   159838             1                  0   
25880002    895835   159838             1                  0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered  \
13645692                  3          0   
17251990                  6          0   
17622767                  2          0   
24138593                  5          0   
25880002                  1          0   

                                               product_name  ...  \
13645692                            Organic Red Bell Pepper  ...   
17251990                               Goat Cheese Crumbles  ...   
17622767                           California Veggie Burger  ...   
24138593                              Organic Peeled Garlic  ...   
25880002  4X Ultra Concentrated Natural Laundry Detergen...  ...   

            price_range_loc  busiest_day  Busiest_days busiest_period_of_day  \
13645692  Mid-range product  Busiest day  Busiest Days        Average orders   
17251990  Mid-range product  Busiest day  Busiest Days        Average orders   
17622767  Low-range product  Busiest day  Busiest Days        Average orders   
24138593  Mid-range product  Busiest day  Busiest Days        Average orders   
25880002  Low-range product  Busiest day  Busiest Days        Average orders   

         max_order  loyalty_flag mean_price spending_flag  \
13645692         1  New customer       7.42   Low spender   
17251990         1  New customer       7.42   Low spender   
17622767         1  New customer       7.42   Low spender   
24138593         1  New customer       7.42   Low spender   
25880002         1  New customer       7.42   Low spender   

          median_days_since_prior_order order_frequency_flag  
13645692                            NaN                  NaN  
17251990                            NaN                  NaN  
17622767                            NaN                  NaN  
24138593                            NaN                  NaN  
25880002                            NaN                  NaN  

[5 rows x 24 columns]

In [34]:
ords_prods_merge.shape

(32404859, 24)

In [35]:
# Export data
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', '4.8_task_ords_prods_merge_grouping data'))